packages/blocks/src/database-block/components/column-type/progress.ts

```diff
@@ -0,0 +1,207 @@
+      this._onPointerDown
+    );
+    disposables.addFromEvent(this, 'pointermove', this._onPointerMove);
+    disposables.addFromEvent(this, 'pointerup', this._onPointerUp);
+    disposables.addFromEvent(document, 'pointermove', this._onDocumentMove);
+
+    if (this.cell?.value) {
+      const { width } = this._progressBg.getBoundingClientRect();
+      const visibleWidth = width - 6;
+      const x = visibleWidth * (this.cell.value / 100);
+      this._translateX = x;
+      this._dragHandle.style.transform = `translate(${x}, -1px)`;
+    }
+  }
+
+  private _onDocumentMove = () => {
+    if (!this._dragConfig) return;
+    this._onPointerUp();
+  };
+
+  private _onPointerDown = (event: PointerEvent) => {
+    event.stopPropagation();
+    const { left, width } = this._progressBg.getBoundingClientRect();
+    const visibleWidth = width - 6;
+    this._dragConfig = {
+      stepWidth: visibleWidth / 100,
+      boundLeft: left,
+      containerWidth: visibleWidth,
+    };
+  };
+
+  private _onPointerMove = (event: PointerEvent) => {
+    event.stopPropagation();
+    if (!this._dragConfig) return;
+    const x = event.clientX;
+    const { boundLeft, containerWidth, stepWidth } = this._dragConfig;
+
+    let steps: number;
+    if (x <= boundLeft) {
+      steps = 0;
+      this._positionX = 0;
+    } else if (x - boundLeft >= containerWidth) {
+      steps = 100;
+      this._positionX = containerWidth;
+    } else {
+      steps = Math.floor((x - boundLeft) / stepWidth);
+      this._positionX = Math.abs(x - boundLeft);
+    }
+
+    if (this.cell?.value !== steps) {
+      this.rowHost.setValue(steps);
+    }
+  };
+
+  private _onPointerUp = () => {
+    this._dragConfig = null;
+  };
+
+  protected override render() {
+    const progress = this.cell?.value ?? 0;
+
+    const fgStyles = styleMap({
+      width: `${progress}%`,
+    });
+
+    const dragStyles = styleMap({
+      transform: `translate(${
+        this._positionX !== -1 ? this._positionX : this._translateX
+      }px, -1px)`,
+    });
+
+    return html`<div
+      class="affine-database-progress"
+      @mousedown=${(e: Event) => e.preventDefault()}
+    >
+      <div class="affine-database-progress-bar">
+        <div class="affine-database-progress-bg">
+          <div class="affine-database-progress-fg" style=${fgStyles}></div>
+          <div
+            class="affine-database-progress-drag-handle"
+            style=${dragStyles}
+          ></div>
+        </div>
+      </div>
+      <div class="progress-number progress">${progress}</div>
+    </div>`;
+  }
+}
+
+@customElement('affine-database-progress-column-property-editing')
+class ProgressColumnPropertyEditing extends DatabaseCellElement<number> {
+  static override tag = literal`affine-database-progress-column-property-editing`;
+}
+
+export const ProgressColumnRenderer = defineColumnRenderer(
+  'progress',
+  () => ({}),
+  () => 0,
+  {
+    Cell: ProgressCellEditing,
+    CellEditing: false,
```

Comment:
I think `null` here makes more sense

Reply:
Currently, all column types depend on this field.

I will open a new pr to modify it, and optimize some redundant fields in the original design.

Reply:
That would be great!

Tips: you can reference this issue comment in your refactoring PR :)